executor/admin.go

```diff
@@ -138,3 +145,319 @@ func (e *CheckIndexRangeExec) constructIndexScanPB() *tipb.Executor {
+	e.recoverRows = e.recoverRows[:0]
+	handleIdx := len(e.columns) - 1
+	result.scanRowCount = 0
+Loop:
+	for {
+		err := srcResult.NextChunk(ctx, e.srcChunk)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		if e.srcChunk.NumRows() == 0 {
+			break
+		}
+		iter := chunk.NewIterator4Chunk(e.srcChunk)
+		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+			if result.scanRowCount >= int64(e.batchSize) {
+				break Loop
+			}
+			handle := row.GetInt64(handleIdx)
+			idxVals := e.extractIdxVals(row, e.idxValsBufs[result.scanRowCount])
+			e.idxValsBufs[result.scanRowCount] = idxVals
+			e.recoverRows = append(e.recoverRows, recoverRows{handle: handle, idxVals: idxVals, skip: false})
+			result.scanRowCount++
+			result.nextHandle = handle + 1
+		}
+	}
+
+	return e.recoverRows, nil
+}
+
+func (e *RecoverIndexExec) batchMarkDup(txn kv.Transaction, rows []recoverRows) error {
+	if len(rows) == 0 {
+		return nil
+	}
+
+	sc := e.ctx.GetSessionVars().StmtCtx
+	batchKeys := make([]kv.Key, 0, len(rows))
+	distinctFlags := make([]bool, len(rows))
+	for i, row := range rows {
+		idxKey, distinct, err := e.index.GenIndexKey(sc, row.idxVals, row.handle, e.idxKeyBufs[i])
+		if err != nil {
+			return errors.Trace(err)
+		}
+		e.idxKeyBufs[i] = idxKey
+
+		batchKeys = append(batchKeys, idxKey)
+		distinctFlags[i] = distinct
+	}
+
+	values, err := txn.GetSnapshot().BatchGet(batchKeys)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// 1. unique-key is duplicate and the handle is equal, skip it.
+	// 2. unique-key is duplicate and the handle is not equal, data is not consistent, log it and skip it.
+	// 3. non-unique-key is duplicate, skip it.
+	for i, key := range batchKeys {
+		if val, found := values[string(key)]; found {
+			if distinctFlags[i] {
+				handle, err1 := tables.DecodeHandle(val)
+				if err1 != nil {
+					return errors.Trace(err1)
+				}
+
+				if handle != rows[i].handle {
+					log.Warnf("[recover-index] The constraint of unique index:%v is broken, handle:%v is not equal handle:%v with idxKey:%v.",
+						e.index.Meta().Name.O, handle, rows[i].handle, key)
+				}
+			}
+			rows[i].skip = true
+		}
+	}
+	return nil
+}
+
+func (e *RecoverIndexExec) backfillIndexInTxn(ctx context.Context, txn kv.Transaction, startHandle int64) (result backfillResult, err error) {
+	result.nextHandle = startHandle
+	srcResult, err := e.buildTableScan(ctx, txn, e.table, startHandle, uint64(e.batchSize))
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	defer terror.Call(srcResult.Close)
+
+	rows, err := e.fetchRecoverRows(ctx, srcResult, &result)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+
+	err = e.batchMarkDup(txn, rows)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+
+	oldBatchCheck := e.ctx.GetSessionVars().StmtCtx.BatchCheck
+	// Constrains is already checked.
+	e.ctx.GetSessionVars().StmtCtx.BatchCheck = true
```

Comment:
Cloud we put it to NextChunk?
And Why we need to change it to oldBatchCheck? It seems `StmtCtx` only affects the current statement.